loki/transformations/block_index_transformations.py

```diff
@@ -0,0 +1,416 @@
+        definitions = []
+        if field_array_mod_imports:
+            definitions += self._create_dummy_field_api_defs(field_array_mod_imports)
+
+        # propagate dummy field_api wrapper definitions to children
+        self.propagate_defs_to_children(self._key, definitions, successors)
+
+        #TODO: we also need to process any code inside a loki/acdc parallel pragma at the driver layer
+
+    def build_ydvars_global_gfl_ptr(self, var):
+        """Replace accesses to thread-local ``YDVARS%GFL_PTR`` with global ``YDVARS%GFL_PTR_G``."""
+
+        if (parent := var.parent):
+            parent = self.build_ydvars_global_gfl_ptr(parent)
+
+        _type = var.type
+        if 'gfl_ptr' in var.basename.lower():
+            _type = parent.variable_map['gfl_ptr_g'].type
+
+        return var.clone(name=var.name.upper().replace('GFL_PTR', 'GFL_PTR_G'),
+                         parent=parent, type=_type)
+
+    def process_body(self, body, definitions, successors, targets, exclude_arrays):
+
+        # build list of type-bound array access using the horizontal index
+        _vars = [var for var in FindVariables(unique=False).visit(body)
+                if isinstance(var, Array) and var.parents and self.horizontal.index in var.dimensions]
+
+        # build list of type-bound view pointers passed as subroutine arguments
+        for call in FindNodes(ir.CallStatement).visit(body):
+            if call.name in targets:
+                _args = {a: d for d, a in call.arg_map.items() if isinstance(d, Array)}
+                _vars += [a for a, d in _args.items()
+                         if any(v in d.shape for v in self.horizontal.size_expressions) and a.parents]
+
+        # replace per-block view pointers with full field pointers
+        vmap = {var: var.clone(name=var.name_parts[-1] + '_FIELD',
+                               type=var.parent.variable_map[var.name_parts[-1] + '_FIELD'].type)
+                for var in _vars}
+
+        # replace thread-private GFL_PTR with global
+        if self.global_gfl_ptr:
+            vmap.update({v: self.build_ydvars_global_gfl_ptr(vmap.get(v, v))
+                         for v in FindVariables(unique=False).visit(body) if 'ydvars%gfl_ptr' in v.name.lower()})
+            vmap = recursive_expression_map_update(vmap)
+
+        # filter out arrays marked for exclusion
+        vmap = {k: v for k, v in vmap.items() if not any(e in k for e in exclude_arrays)}
+
+        # propagate dummy field_api wrapper definitions to children
+        self.propagate_defs_to_children(self._key, definitions, successors)
+
+        # finally we perform the substitution
+        return SubstituteExpressions(vmap).visit(body)
+
+
+    def process_kernel(self, routine, item, successors, targets, exclude_arrays):
+
+        # Sanitize the subroutine
+        resolve_associates(routine)
+        v_index = SCCBaseTransformation.get_integer_variable(routine, name=self.horizontal.index)
```

Comment:
[no action] Note to self, these should be separated out into general utilities. The fact that this field trafo is calling into the SCC stack signals the problem.